### PR TITLE
Build with cmake

### PR DESCRIPTION
**Pull request by jschlight**
I've created a couple of example projects that show how to build N-API addons using CMake.js. Once this PR lands, I plan to add links to it, along with some narrative, to https://github.com/nodejs/abi-stable-node/issues/381 and https://github.com/nodejs/node-addon-api/issues/445.

### CODE REVIEW

File: build_with_cmake/node-addon-api/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+file(GLOB SOURCE_FILES "hello.cc")
+add_library(${PROJECT_NAME} SHARED ${SOURCE_FILES} ${CMAKE_JS_SRC})
+set_target_properties(${PROJECT_NAME} PROPERTIES PREFIX "" SUFFIX ".node")
+target_link_libraries(${PROJECT_NAME} ${CMAKE_JS_LIB})
+
+# define NPI_VERSION
+add_definitions(-DNAPI_VERSION=2)
+
+# Include N-API wrappers
+target_include_directories(${PROJECT_NAME} PRIVATE "${CMAKE_SOURCE_DIR}/node_modules/node-addon-api")
```
Comment (NickNaso) [inline build_with_cmake/node-addon-api/CMakeLists.txt]: Could be possible call:
`node -p "require('node-addon-api').include" ` that return the absolute path for your include dir  instead to set it by hand?

File: build_with_cmake/napi/package.json [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "hello_world",
+  "version": "0.0.0",
+  "description": "Node.js Addons Example #1",
```
[the remark this answers is not in the document]
Comment (jschlight) [reply build_with_cmake/napi/package.json]: Fixed.

File: build_with_cmake/napi/package.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "hello_world",
```
[the remark this answers is not in the document]
Comment (jschlight) [reply build_with_cmake/napi/package.json]: Fixed.

File: build_with_cmake/node-addon-api/package.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
+  "name": "hello_world",
+  "version": "0.0.0",
+  "description": "Node.js Addons Example #1",
```
[the remark this answers is not in the document]
Comment (jschlight) [reply build_with_cmake/node-addon-api/package.json]: Fixed.

File: build_with_cmake/node-addon-api/package.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
+{
+  "name": "hello_world",
```
[the remark this answers is not in the document]
Comment (jschlight) [reply build_with_cmake/node-addon-api/package.json]: Fixed.

File: build_with_cmake/node-addon-api/CMakeLists.txt (hunk 1 above)
Comment (jschlight) [reply build_with_cmake/node-addon-api/CMakeLists.txt]: Good find, Nicola. 

I've updated `CMakeLists.txt` to use the technique recommended by the [CMake.js team](https://github.com/cmake-js/cmake-js#n-api-and-node-addon-api).

This technique was failing for me before, but now it's working again. It's probably due to the three lines I added earlier to the very top of `CMakeLists.txt `.

File: build_with_cmake/napi/hello.cc [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#include <assert.h>
+#include <node_api.h>
+
+napi_value Method(napi_env env, napi_callback_info info) {
```
Comment (gabrielschulhof) [inline build_with_cmake/napi/hello.cc]: This method should be static.

File: build_with_cmake/napi/hello.cc [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  napi_value world;
+  status = napi_create_string_utf8(env, "Hello, world!", 13, &world);
+  assert(status == napi_ok);
+  return world;
+}
+
+#define DECLARE_NAPI_METHOD(name, func)                                        \
+  { name, 0, func, 0, 0, 0, napi_default, 0 }
+
+napi_value Init(napi_env env, napi_value exports) {
```
Comment (gabrielschulhof) [inline build_with_cmake/napi/hello.cc]: This method should also be static.

File: build_with_cmake/node-addon-api/hello.cc [hunk 8]
```diff
@@ -0,0 +1,14 @@
+#include <napi.h>
+
+Napi::String Method(const Napi::CallbackInfo& info) {
```
Comment (gabrielschulhof) [inline build_with_cmake/node-addon-api/hello.cc]: This method should be static.

File: build_with_cmake/node-addon-api/hello.cc [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  Napi::Env env = info.Env();
+  return Napi::String::New(env, "Hello, world!");
+}
+
+Napi::Object Init(Napi::Env env, Napi::Object exports) {
```
Comment (gabrielschulhof) [inline build_with_cmake/node-addon-api/hello.cc]: Ditto.

**Comment by jschlight** [comment]
PR merged as discussed during today's N-API team weekly meeting.

**Comment by jschlight** [comment]
Arrg! I'm creating a new PR to implement Gabriel's requests. It should be up momentarily.

**Comment by jschlight** [comment]
I am tracking down why the identical build that succeeded three days ago is now failing. Once I have that issue resolved, I'll create a new PR that fixes the build and adds Gabriel's requests. 

2019-09-20 https://travis-ci.org/jschlight/node-addon-examples/builds/587656587
2019-09-23 https://travis-ci.com/nodejs/node-addon-examples/builds/128799292

File: build_with_cmake/node-addon-api/hello.cc (hunk 8 above)
Comment (jschlight) [reply build_with_cmake/node-addon-api/hello.cc]: Fixed in https://github.com/nodejs/node-addon-examples/pull/114.

File: build_with_cmake/napi/hello.cc (hunk 7 above)
Comment (jschlight) [reply build_with_cmake/napi/hello.cc]: Fixed in https://github.com/nodejs/node-addon-examples/pull/114.

File: build_with_cmake/napi/hello.cc (hunk 6 above)
Comment (jschlight) [reply build_with_cmake/napi/hello.cc]: Fixed in https://github.com/nodejs/node-addon-examples/pull/114.

File: build_with_cmake/node-addon-api/hello.cc (hunk 9 above)
Comment (jschlight) [reply build_with_cmake/node-addon-api/hello.cc]: Fixed in https://github.com/nodejs/node-addon-examples/pull/114.

**Pull request merged**